### DISCUSSION TITLE
Why isn't lua_ls able to get annotations straight from $VIMRUNTIME?

### DISCUSSION THREAD

**Discussion by d-r-a-b** [folke/lua-dev.nvim#146] [Q&A] [answered]
I was trying to get completion and documentation (`hover()`) setup with `nvim-lspconfig` for neovim and the vim.* namespace. Research led me to `neodev.nvim` and I saw that `https://github.com/folke/neodev.nvim/blob/b1572ef7d09cb324bf0e284c3e363b1f55193e6b/.github/workflows/types.yml` is basically pulling annotations from the `neovim` codebase and then the plugin is somehow surfacing those annotations. Why this seemingly roundabout approach? I had expected that setting `workspace.library` to `vim.api.nvim_get_runtime_file("", true)` would have been enough to get this info from `$VIMRUNTIME` but I was unable to get this to work. I'm sure there's a good reason but I'm trying to understand it and thereby expand my understanding of neovim, neovim plugins and LSPs.

Context:
> I've been using vim/neovim without plugins and as a comparatively basic editor for a while now and it worked well for my own (usually 1-3 file) projects. Recently, I've been trying to work my way through neovim plugin codebases and felt that having LSP features would help to be able to get documentation and jump around the codebase so I started trying to set that up with folke's excellent `lazy.nvim`. I'm new to setting up plugins in general and also to `lazy.nvim` specifically, but things have overall been going well and I am just trying to flesh out my knowledge on my journey from neovim plugin/LSP neophyte to neovim plugin/LSP intermediate learner.
>
> I copied the config block from [lspconfig wiki](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#lua_ls), I set `workspace.library` to `vim.api.nvim_get_runtime_file("", true)` . Based on the comment "Make the server aware of Neovim runtime files" I would have thought this would work to get completions and `hover()` docs going. It does at least give me some completion with `nvim-cmp-lsp` but no documentation on `vim.lsp.buf.hover()`. Was this supposed to work?
>
> Searching around brought me again to folke's `neodev.nvim` and this community. In general, I like to understand what my tools are doing for me and subscribe to a philosophy of reducing dependencies when possible. At the moment, `neodev.nvim` makes things work when before they didn't, so it's obviously a good value proposition to use it. However, it feels like `neovim` and `lua-language-server` should be able to do this without needing yet another plugin, so I wanted to understand better what brilliant things `neodev.nvim` might be doing to glue everything together. That led me to the `types` workflow that I linked above, and hence my question.

Thank you in advance to anyone who is able and willing to take the time to provide the benefit of their knowledge and experience, and thank you folke for providing this and many other wonderful plugins.

**Comment by folke** [comment, marked as answer]
Because the Neovim code-base is not just lua? Everything under `vim.api.nvim_` is all c. lua_ls does't understand any of that

**Reply by d-r-a-b** [reply]
d'oh! Ignorance is a terrible thing. Thank you very much for taking the time to reply and helping me push back some of mine.

For whatever reason I just thought the core library would have been written in Lua, even though if I take a step back and really think about the evolution of neovim from vim that makes no sense. Trying to keep a bunch on things in my mental state right now to get some things working, and just completely didn't consider it.

Thanks again!
